local-interchaintest/src/setup.rs

```diff
@@ -0,0 +1,548 @@
+        ctx.build_tx_update_auction_oracle().send()?;
+
+        util::create_deployment_file(
+            ctx.get_astroport_factory()?
+                .contract_addr
+                .expect("missing deployed astroport factory")
+                .as_str(),
+            ctx.get_auctions_manager()?
+                .contract_addr
+                .expect("missing deployed astroport factory")
+                .as_str(),
+        )?;
+        util::create_arbs_file()?;
+        util::create_netconfig()?;
+        util::create_denom_file(&self.denom_map)?;
+
+        let statuses = self.test_statuses.clone();
+
+        test.setup(&mut self.denom_map, self.test_ctx)?;
+
+        with_arb_bot_output(Arc::new(Box::new(move |arbfile: Value| {
+            statuses.lock().expect("Failed to lock statuses").insert(
+                (test.name.clone(), test.description.clone()),
+                (*test.test)(arbfile),
+            );
+
+            Ok(())
+        })))?;
+
+        Ok(self)
+    }
+
+    /// Produces a single result representing any failure that may
+    /// have occurred in any test.
+    /// Logs successes to stdout, and failures to stderr.
+    pub fn join(&mut self) -> TestResult {
+        for ((name, description), status) in self
+            .test_statuses
+            .lock()
+            .expect("Failed to lock test statuses")
+            .iter()
+        {
+            match status {
+                Ok(_) => {
+                    println!("✅ 🤑 SUCCESS {name} - {description}");
+                }
+                Err(e) => {
+                    eprintln!("❌ ඞ FAILURE {name} - {description}: {:?}", e);
+                }
+            }
+        }
+
+        if let Some(e) = self
+            .test_statuses
+            .lock()
+            .expect("Failed to lock test statuses")
+            .drain()
+            .filter_map(|res| res.1.err())
+            .next()
+        {
+            return Err(e);
+        }
+
+        Ok(())
+    }
+}
+
+/// A test that receives arb bot executable output.
+pub type TestFn = Box<dyn Fn(Value) -> TestResult + Send + Sync>;
+pub type OwnedTestFn = Arc<TestFn>;
+pub type TestResult = Result<(), Box<dyn Error + Send + Sync>>;
+
+/// Defines a test case. A test case is characterized by its
+/// pools, and the balances associated with each pool,
+/// and its auctions, and the price associated with each auction.
+#[derive(Builder)]
+#[builder(setter(into, strip_option, prefix = "with"), pattern = "owned")]
+pub struct Test {
+    /// Test metadata
+    name: String,
+    description: String,
+
+    /// Fully qualified denoms (i.e., factory/neutronxyz/tokenabc or untrn or ibc/xyz)
+    denoms: Vec<String>,
+
+    /// How much of a given subdenom acc0 owns on a given chain
+    /// (chain, token) -> balance
+    tokenfactory_token_balances_acc0: HashMap<(String, String), u128>,
+
+    /// (Denom a, denom b) or (offer asset, ask asset) -> pool
+    pools: HashMap<(String, String), Vec<Pool>>,
```

Comment:
do we have a vec of pools here because some pairs may have multiple types of pools?